File: backend-rust/src/external/task.rs
use std::sync::Arc;

use axum::extract::{Query, State};
use axum::http::HeaderMap;
use axum::response::IntoResponse;
use axum::Json;
use serde::Deserialize;
use serde_json::{json, Value};

use crate::db;
use crate::error::RestError;
use crate::model::Task;

use super::{authenticate, cursor, ExternalState};

/// backend(Go)のtaskDTOToJSON(internal/handler/external/task.go)と同じ形状。
/// 内部CRUDのレスポンスとは異なり**user_idを含めない**点に注意
fn task_to_json(task: &Task) -> Value {
    let labels: Vec<Value> = task
        .labels
        .iter()
        .map(|l| json!({"id": l.id, "name": l.name}))
        .collect();
    json!({
        "id": task.id,
        "name": task.name,
        "description": task.description,
        "status": task.status.as_str(),
        "finished_on": task.finished_on.format("%Y-%m-%d").to_string(),
        "labels": labels,
        "created_at": task.created_at.and_utc().to_rfc3339(),
        "updated_at": task.updated_at.and_utc().to_rfc3339(),
    })
}

#[derive(Debug, Deserialize)]
pub struct ListQuery {
    user_id: Option<String>,
    // v1(offset)
    page: Option<i64>,
    page_size: Option<i64>,
    // v2(cursor)
    cursor: Option<String>,
    limit: Option<i64>,
}

/// GET /external/v1/tasks
/// `backend.external-tasks-pagination-v2`(5言語で共有する1つのFeature Flag)のON/OFFで
/// offsetページング(v1)/keysetページング(v2)を切り替える(backend(Go)のTaskHandler.Listと同じ)
pub async fn list(
    State(state): State<Arc<ExternalState>>,
    headers: HeaderMap,
    Query(q): Query<ListQuery>,
) -> Result<impl IntoResponse, RestError> {
    authenticate(&state, &headers).await?;

    let user_id_str = q.user_id.ok_or(RestError::UserIdRequired)?;
    if user_id_str.is_empty() {
        return Err(RestError::UserIdRequired);
    }
    let user_id: u64 = user_id_str.parse().map_err(|_| RestError::InvalidUserId)?;

    let use_v2 = state.pagination_v2.get();
    if use_v2 {
        list_v2(&state, user_id, q.cursor, q.limit).await
    } else {
        list_v1(&state, user_id, q.page, q.page_size).await
    }
}

async fn list_v1(
    state: &ExternalState,
    user_id: u64,
    page: Option<i64>,
    page_size: Option<i64>,
) -> Result<Json<Value>, RestError> {
    let page = page.filter(|p| *p >= 1).unwrap_or(1);
    let page_size = page_size.filter(|p| *p >= 1).unwrap_or(10);

    let (tasks, total) = db::list_tasks_offset_external(&state.pool, user_id, page, page_size)
        .await
        .map_err(|_| RestError::Internal)?;

    let tasks_json: Vec<Value> = tasks.iter().map(task_to_json).collect();
    Ok(Json(json!({
        "tasks": tasks_json,
        "page": page,
        "page_size": page_size,
        "total": total,
    })))
}

async fn list_v2(
    state: &ExternalState,
    user_id: u64,
    cursor_str: Option<String>,
    limit: Option<i64>,
) -> Result<Json<Value>, RestError> {
    let limit = limit.filter(|l| *l >= 1).unwrap_or(10);

    let after = match cursor_str.filter(|c| !c.is_empty()) {
        Some(c) => Some(cursor::decode(&c).map_err(RestError::InvalidCursor)?),
        None => None,
    };

    let tasks = db::list_tasks_cursor_external(&state.pool, user_id, after, limit)
        .await
        .map_err(|_| RestError::Internal)?;

    let next_cursor: Value = if tasks.len() as i64 == limit {
        if let Some(last) = tasks.last() {
            json!(cursor::encode(last.created_at, last.id))
        } else {
            Value::Null
        }
    } else {
        Value::Null
    };

    let tasks_json: Vec<Value> = tasks.iter().map(task_to_json).collect();
    Ok(Json(json!({
        "tasks": tasks_json,
        "next_cursor": next_cursor,
        "limit": limit,
    })))
}

File: backend-js-express/src/external/task.js
'use strict';

const db = require('../db');
const errors = require('../error');
const cursorMod = require('./cursor');
const { authenticate } = require('./authenticate');
const { statusToString } = require('../model');
const { mysqlDatetimeToRfc3339, isoToMysqlDatetime } = require('../time');

function wrap(fn) {
  return (req, res, next) => Promise.resolve(fn(req, res)).catch(next);
}

// backend(Go)のtaskDTOToJSON(internal/handler/external/task.go)と同じ形状。
// 内部CRUDのレスポンスと同様に**user_idを含めない**点に注意
function taskToJson(task) {
  return {
    id: task.id,
    name: task.name,
    description: task.description,
    status: statusToString(task.status),
    finished_on: task.finishedOn,
    labels: task.labels.map((l) => ({ id: l.id, name: l.name })),
    created_at: mysqlDatetimeToRfc3339(task.createdAt),
    updated_at: mysqlDatetimeToRfc3339(task.updatedAt),
  };
}

// GET /external/v1/tasks
// `backend.external-tasks-pagination-v2`(全言語で共有する1つのFeature Flag)のON/OFFで
// offsetページング(v1)/keysetページング(v2)を切り替える(backend(Go)のTaskHandler.Listと同じ)
function list(state) {
  return wrap(async (req, res) => {
    await authenticate(state, req);

    const userIdStr = req.query.user_id;
    if (!userIdStr) throw errors.userIdRequired();
    if (!/^\d+$/.test(userIdStr)) throw errors.invalidUserId();
    const userId = Number(userIdStr);

    if (state.paginationV2Flag.get()) {
      await listV2(state, userId, req.query, res);
    } else {
      await listV1(state, userId, req.query, res);
    }
  });
}

async function listV1(state, userId, q, res) {
  const page = q.page && Number(q.page) >= 1 ? Number(q.page) : 1;
  const pageSize = q.page_size && Number(q.page_size) >= 1 ? Number(q.page_size) : 10;

  const { tasks, total } = await db.listTasksOffsetExternal(state.pool, userId, page, pageSize);
  res.json({ tasks: tasks.map(taskToJson), page, page_size: pageSize, total });
}

async function listV2(state, userId, q, res) {
  const limit = q.limit && Number(q.limit) >= 1 ? Number(q.limit) : 10;

  let after = null;
  if (q.cursor) {
    let decoded;
    try {
      decoded = cursorMod.decode(q.cursor);
    } catch (msg) {
      throw errors.invalidCursor(String(msg));
    }
    const mysqlDt = isoToMysqlDatetime(decoded.createdAtIso);
    if (!mysqlDt) throw errors.invalidCursor('created_atの形式が不正です');
    after = { createdAt: mysqlDt, id: decoded.id };
  }

  const tasks = await db.listTasksCursorExternal(state.pool, userId, after, limit);

  let nextCursor = null;
  if (tasks.length === limit) {
    const last = tasks[tasks.length - 1];
    nextCursor = cursorMod.encode(mysqlDatetimeToRfc3339(last.createdAt), last.id);
  }
  res.json({ tasks: tasks.map(taskToJson), next_cursor: nextCursor, limit });
}

module.exports = { list, taskToJson };
